utils/appliance/implementations/ui.py

```diff
@@ -374,14 +374,19 @@ def appliance(self):
     def widgetastic(self):
         """This gives us a widgetastic browser."""
         # TODO: Make this a property that could watch for browser change?
-        return MiqBrowser(self.open_browser(), self)
+        try:
+            return MiqBrowser(self.open_browser(), self)
+        finally:
+            manager.add_cleanup(self._reset_cache)
```

Comment:
This doesn't seem right still to me. You mentioned that the clean up is added per browser, because it's tied to the browser, whilst I understand this and now agree with you, this code is not seemingly doing that. It's adding the same function, every time the widgetastic is created. So it will just call the reset_cache function potentially multiple times.